### PR TITLE
cuco::bloom_filter

### PR DESCRIPTION
**Pull request by sleeepyjack**
Adds a new class called `cuco::bloom_filter` for approximate set membership queries.

It is used to test whether an element is a member of a set. False positive matches are possible, but false negatives are not – in other words, a query returns either "possibly in set" or "definitely not in set". Elements can be added to the set, but not removed; the more items added, the larger the probability of false positives.

The type of implementation used here is known as a "partitioned" or "pattern-blocked" bloom filter.

This PR comes with examples, benchmarks, as well as unit tests.

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by sleeepyjack** [comment]
ok to test

**Comment by sleeepyjack** [comment]
Meh, forgot that I don't have the permissions to fire up the CI.

This PR is ready to test and ready for review.

**Comment by jrhemstad** [comment]
add to whitelist

**Comment by jrhemstad** [comment]
okay to test

**Comment by jrhemstad** [comment]
ok to test

**Comment by dillon-cullinan** [comment]
add to whitelist

**Comment by sleeepyjack** [comment]
ok to test

**Comment by sleeepyjack** [comment]
rerun tests

**Comment by jrhemstad** [comment]
rerun tests

File: include/cuco/detail/bloom_filter_kernels.cuh [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+ * @tparam View Type of device view allowing access of filter storage
+ * @tparam Hash Unary callable type
+ * @param first Beginning of the sequence of keys
+ * @param last End of the sequence of keys
+ * @param output_begin Beginning of the sequence of booleans for the presence of each key
+ * @param view Mutable device view used to access the filter's slot storage
+ * @param hash The unary function to apply to hash each key
+ */
+template <std::size_t block_size, typename InputIt, typename OutputIt, typename View, typename Hash>
+__global__ void contains(InputIt first, InputIt last, OutputIt output_begin, View view, Hash hash)
```
Comment (gevtushenko) [inline include/cuco/detail/bloom_filter_kernels.cuh]: Specifying launch bounds sometimes affects performance:
```
__global__ void __launch_bounds__(block_size)
```

File: include/cuco/detail/bloom_filter_kernels.cuh [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    /*
+     * The ld.relaxed.gpu instruction used in view.contains causes L1 to
+     * flush more frequently, causing increased sector stores from L2 to global memory.
+     * By writing results to shared memory and then synchronizing before writing back
+     * to global, we no longer rely on L1, preventing the increase in sector stores from
+     * L2 to global and improving performance.
+     */
+    typename View::key_type const key{*(first + tid)};
+    write_buffer[threadIdx.x] = view.contains(key, hash);
+    __syncthreads();
```
Comment (gevtushenko) [inline include/cuco/detail/bloom_filter_kernels.cuh]: Interesting optimization! As far as I understand, it's unsafe to synchronize like this. If input size isn't a multiple of a `block_size`, `__syncthreads()` won't be called by all threads of a thread block. Or [this  information](https://stackoverflow.com/questions/6666382/can-i-use-syncthreads-after-having-dropped-threads) is outdated and it's fine :smile:

File: include/cuco/detail/bloom_filter_kernels.cuh [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+/**
+ * @brief Indicates whether the keys in the range `[first, last)` are contained
+ * in the filter.
+ *
+ * Writes a `bool` to `(output + i)` indicating if the key `*(first + i)` exists
+ * in the filter.
+ *
+ * @tparam block_size The size of the thread block
+ * @tparam InputIt Device accessible input iterator whose `value_type` is
```
Comment (gevtushenko) [inline include/cuco/detail/bloom_filter_kernels.cuh:84]: As I understand input iterators don't enforce `equality_comparable` property (unlike legacy input iterators or random access iterators). If I'm not mistaken, we might need to rewrite `(first + tid) < last` as `auto size = distance(first, last); tid < size` or require legacy input iterators in the documentation. I'm not particularly strong in the field of iterator concepts, so correct me if I'm wrong :sweat_smile:

File: include/cuco/detail/bloom_filter_kernels.cuh (hunk 2 above)
Comment (jrhemstad) [reply include/cuco/detail/bloom_filter_kernels.cuh]: This originally came from the `static_map`. @sleeepyjack did you test if this optimization actually made a difference on the `bloom_filter` too? But yes, we shouldn't be calling `__syncthreads` in a situation where not all threads will call it.

Comment (PointKernel) [reply include/cuco/detail/bloom_filter_kernels.cuh]: Right, calling `__synchthreads` here is theoretically unsafe. I noticed this issue when implementing multimap but weird enough, compute sanitizer didn't throw an error in this case.

**Comment by jrhemstad** [comment]
@sleeepyjack can you resolve conflicts?

**Comment by sleeepyjack** [comment]
> @sleeepyjack can you resolve conflicts?

on it

File: include/cuco/detail/bloom_filter_kernels.cuh (hunk 2 above)
Comment (sleeepyjack) [reply include/cuco/detail/bloom_filter_kernels.cuh]: I just ran the benchmarks again. There is virtually no performance advantage with the write buffer so I removed it.

File: include/cuco/detail/bloom_filter_kernels.cuh (hunk 1 above)
Comment (sleeepyjack) [reply include/cuco/detail/bloom_filter_kernels.cuh]: I just tested this but sadly there is no performance improvement. Maybe the compiler is already smart enough to deduce the bound from the `block_size` tparam. Or we are simply not limited by occupancy. I need to run a profile to find out.

**Comment by sleeepyjack** [comment]
Now revisiting this PR and giving it some love so we can get it merged.

File: include/cuco/detail/bloom_filter_kernels.cuh (hunk 2 above)
Comment (jrhemstad) [reply include/cuco/detail/bloom_filter_kernels.cuh]: I'd make sure to test on both V100 and A100.

File: include/cuco/detail/bloom_filter_kernels.cuh (hunk 1 above)
Comment (jrhemstad) [reply include/cuco/detail/bloom_filter_kernels.cuh]: It's often just "good practice" to specify the `__launch_bounds__` when the CTA size is known statically. Even if it doesn't impact perf on the benchmarks we have on the specific hardware you tested, there may still be situations (present or future) where it does help.

File: include/cuco/detail/bloom_filter_kernels.cuh (hunk 2 above)
Comment (sleeepyjack) [reply include/cuco/detail/bloom_filter_kernels.cuh]: ['../baseline_v100.json', '../direct_write_v100.json']

## [0] Tesla V100-PCIE-32GB

|  KeyType  |  SlotType  |  Operation  |  FilterScope  |  DataScope  |  NumInputs  |   NumBits    |  NumHashes  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |       Diff |   %Diff |  Status  |
|-----------|------------|-------------|---------------|-------------|-------------|--------------|-------------|------------|-------------|------------|-------------|------------|---------|----------|
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  1000000000  |      2      | 461.451 ms |       0.02% | 461.523 ms |       0.02% |  72.063 us |   0.02% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  1000000000  |      2      |  46.441 ms |       0.52% |  46.450 ms |       0.53% |   9.029 us |   0.02% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  2000000000  |      2      | 480.755 ms |       0.01% | 480.731 ms |       0.01% | -23.465 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  2000000000  |      2      |  48.724 ms |       0.78% |  48.721 ms |       0.77% |  -2.234 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  3000000000  |      2      | 487.495 ms |       0.02% | 487.435 ms |       0.01% | -60.039 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  3000000000  |      2      |  49.731 ms |       1.04% |  49.712 ms |       0.91% | -18.618 us |  -0.04% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  4000000000  |      2      | 490.764 ms |       0.01% | 490.769 ms |       0.01% |   5.027 us |   0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  4000000000  |      2      |  50.428 ms |       1.40% |  50.410 ms |       1.44% | -17.408 us |  -0.03% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  5000000000  |      2      | 493.323 ms |       0.01% | 493.273 ms |       0.01% | -49.710 us |  -0.01% |   FAIL   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  5000000000  |      2      |  51.001 ms |       1.71% |  50.988 ms |       1.70% | -12.939 us |  -0.03% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 10000000000  |      2      | 498.809 ms |       0.01% | 498.773 ms |       0.01% | -36.122 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 10000000000  |      2      |  53.243 ms |       1.55% |  53.221 ms |       1.53% | -21.411 us |  -0.04% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 50000000000  |      2      | 514.476 ms |       0.01% | 514.521 ms |       0.01% |  44.317 us |   0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 50000000000  |      2      |  54.745 ms |       0.08% |  54.755 ms |       0.08% |   9.775 us |   0.02% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 100000000000 |      2      | 499.058 ms |       0.02% | 499.096 ms |       0.01% |  37.983 us |   0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 100000000000 |      2      |  52.598 ms |       4.90% |  52.593 ms |       4.86% |  -5.492 us |  -0.01% |   PASS   |

edit: output shortened - too much to scroll through

Comment (sleeepyjack) [reply include/cuco/detail/bloom_filter_kernels.cuh]: ['../baseline_a100.json', '../direct_write_a100.json']

## [0] NVIDIA A100 80GB PCIe

|  KeyType  |  SlotType  |  Operation  |  FilterScope  |  DataScope  |  NumInputs  |   NumBits    |  NumHashes  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |       Diff |   %Diff |  Status  |
|-----------|------------|-------------|---------------|-------------|-------------|--------------|-------------|------------|-------------|------------|-------------|------------|---------|----------|
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  1000000000  |      2      | 154.549 ms |       0.37% | 154.551 ms |       0.36% |   1.953 us |   0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  1000000000  |      2      |  15.578 ms |       0.99% |  15.577 ms |       1.02% |  -0.372 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  2000000000  |      2      | 171.445 ms |       0.26% | 171.446 ms |       0.26% |   0.562 us |   0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  2000000000  |      2      |  17.357 ms |       0.85% |  17.356 ms |       0.85% |  -0.247 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  3000000000  |      2      | 177.142 ms |       0.24% | 177.137 ms |       0.24% |  -5.493 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  3000000000  |      2      |  18.009 ms |       0.68% |  18.007 ms |       0.68% |  -1.353 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  4000000000  |      2      | 180.017 ms |       0.24% | 180.027 ms |       0.23% |  10.799 us |   0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  4000000000  |      2      |  18.380 ms |       0.47% |  18.378 ms |       0.48% |  -1.536 us |  -0.01% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  |  5000000000  |      2      | 181.806 ms |       0.24% | 181.811 ms |       0.24% |   5.496 us |   0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  |  5000000000  |      2      |  18.631 ms |       0.13% |  18.630 ms |       0.13% |  -0.228 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 10000000000  |      2      | 185.644 ms |       0.29% | 185.638 ms |       0.29% |  -5.862 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 10000000000  |      2      |  18.976 ms |       0.03% |  18.977 ms |       0.03% |   0.721 us |   0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 50000000000  |      2      | 192.314 ms |       0.14% | 192.312 ms |       0.14% |  -1.953 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 50000000000  |      2      |  19.248 ms |       0.03% |  19.252 ms |       0.02% |   4.293 us |   0.02% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     | 1000000000  | 100000000000 |      2      | 192.727 ms |       0.01% | 192.720 ms |       0.01% |  -6.610 us |  -0.00% |   PASS   |
|    I32    |    I32     |   INSERT    |     GMEM      |    GMEM     |  100000000  | 100000000000 |      2      |  19.279 ms |       0.03% |  19.279 ms |       0.03% |   0.000 us |   0.00% |   PASS   |

**Comment by sleeepyjack** [comment]
what whaaat? The ubuntu18.04 runner is complaining. Investigating...
```

/workspace/benchmarks/bloom_filter/bloom_filter_bench.cu:107:1: note: candidate: template<class Key, class Slot, filter_op Op, filter_scope FilterScope, data_scope DataScope> void nvbench_cuco_bloom_filter(nvbench::state&, nvbench::type_list<Key, Slot, nvbench::enum_type<Op, decltype (Op)>, nvbench::enum_type<FilterScope, decltype (FilterScope)>, nvbench::enum_type<DataScope, decltype (DataScope)> >)
 void nvbench_cuco_bloom_filter(nvbench::state& state,
 ^~~~~~~~~~~~~~~~~~~~~~~~~
/workspace/benchmarks/bloom_filter/bloom_filter_bench.cu:107:1: note:   template argument deduction/substitution failed:
/workspace/benchmarks/bloom_filter/bloom_filter_bench.cu:277:163: note:   mismatched types ‘data_scope’ and ‘filter_op’
 NVBENCH_BENCH_TYPES(nvbench_cuco_bloom_filter,
```

**Comment by sleeepyjack** [comment]
I tinkered a bit more with different bit-pattern generators with the goal of improving the FPR and performance (there's obviously a trade-off between the two).

Here are the results:
- plain double hashing
  `h1(k) + i * h2(k)`
  cost: evaluate two hash functions + N times the second term
  FPR: 0.0264

- triple hashing
  `h1(k) + i * h2(k) + i*i * h3(k)`
  cost: evaluate three hash functions + N times the last two terms
  FPR: 0.0204

- extended double hashing
  `h1(k) + i * h2(k) + ((i*i*i - i)/6)`
  cost: evaluate two hash functions + N times the last two terms
  FPR: 0.0236

- best FPR approach
  `h1(k) + h2(k + i)`
  cost: evaluate h1 once and h2 N times (very expensive for long keys)
  FPR: 0.0194
  restriction: operator+ for the term k+i must be available

Given these results, I think I will go with the extended double hashing or triple hashing approach then.

**Comment by sleeepyjack** [comment]
Bildschirmfoto 2022-07-29 um 14 21 22

Wider `Slot` types result in a better FPR. However, since `cuda::atomic<__int128_t>::is_lock_free() == false`, the query throughput drops drastically.

File: include/cuco/detail/bloom_filter.inl [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                                                        Allocator const& alloc)
+  : num_bits_{SDIV(std::max(std::size_t{1}, num_bits), detail::type_bits<slot_type>()) *
+              detail::type_bits<slot_type>()},
+    num_slots_{SDIV(std::max(std::size_t{1}, num_bits), detail::type_bits<slot_type>())},
+    num_hashes_{std::clamp(num_hashes, std::size_t{1}, detail::type_bits<slot_type>())},
+    slot_allocator_{alloc}
+{
+  slots_ = std::allocator_traits<slot_allocator_type>::allocate(slot_allocator_, num_slots_);
+
+  initialize();
```
Comment (PointKernel) [inline include/cuco/detail/bloom_filter.inl]:
```suggestion
  initialize(stream);
```
Minor correction: this ctor should take a stream parameter and use it to initialize the map.

File: include/cuco/detail/bloom_filter.inl [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+#include <cstddef>
+#include <iterator>
+#include <memory>
+
+namespace cuco {
+
+template <typename Key, cuda::thread_scope Scope, typename Allocator, typename Slot>
+bloom_filter<Key, Scope, Allocator, Slot>::bloom_filter(std::size_t num_bits,
+                                                        std::size_t num_hashes,
+                                                        Allocator const& alloc)
```
Comment (PointKernel) [inline include/cuco/detail/bloom_filter.inl]:
```suggestion
                                                        Allocator const& alloc
                                                        cudaStream_t stream)
```

**Comment by sleeepyjack** [comment]
I'm dropping the `cuda::annotated_ptr`/`cuda::apply_access_policy` strategy as the access policy is apparently not applied correctly (virtual no performance difference between L2-persistent and non-persistent filters). Thus, I'm rolling back to the old strategy, i.e., using the CUDA driver API.

Here are some benchmark results on A100 80GB L2-resident vs. non-resident filter:
| KeyType | SlotType | FilterOperation | FilterScope | DataScope | NumInputs |  NumBits  | NumHashes | nv/filter/fpr | nv/filter/size/mb | Samples |  CPU Time  | Noise |  GPU Time  | Noise | Elem/s  | GlobalMem BW | BWUtil | Samples | Batch GPU  |
|---------|----------|-----------------|-------------|-----------|-----------|-----------|-----------|---------------|-------------------|---------|------------|-------|------------|-------|---------|--------------|--------|---------|------------|
|     I32 |      U64 |          INSERT |        GMEM |      GMEM |  10000000 | 300000000 |         2 |     0.0059597 |                37 |    773x | 656.308 us | 1.37% | 647.584 us | 0.22% | 15.442G | 123.536 GB/s |  6.38% |    814x | 641.942 us |
|     I32 |      U64 |          INSERT |        GMEM |      GMEM | 100000000 | 300000000 |         2 |       0.24194 |                37 |     81x |   6.194 ms | 1.50% |   6.185 ms | 1.49% | 16.168G | 129.345 GB/s |  6.68% |     85x |   6.163 ms |
|     I32 |      U64 |          INSERT |        GMEM |      REGS |  10000000 | 300000000 |         2 |     0.0059597 |                37 |   1289x | 396.613 us | 2.47% | 387.908 us | 1.02% | 25.779G | 206.235 GB/s | 10.66% |   1374x | 380.201 us |
|     I32 |      U64 |          INSERT |        GMEM |      REGS | 100000000 | 300000000 |         2 |       0.24194 |                37 |    171x |   2.940 ms | 0.87% |   2.932 ms | 0.81% | 34.111G | 272.888 GB/s | 14.10% |    180x |   2.940 ms |
|     I32 |      U64 |          INSERT |          L2 |      GMEM |  10000000 | 300000000 |         2 |     0.0059597 |                37 |   1819x | 283.593 us | 3.21% | 274.877 us | 0.51% | 36.380G | 291.039 GB/s | 15.04% |   1896x | 269.990 us |
|     I32 |      U64 |          INSERT |          L2 |      GMEM | 100000000 | 300000000 |         2 |       0.24194 |                37 |    201x |   2.505 ms | 0.74% |   2.496 ms | 0.64% | 40.060G | 320.483 GB/s | 16.56% |    202x |   2.519 ms |
|     I32 |      U64 |          INSERT |          L2 |      REGS |  10000000 | 300000000 |         2 |     0.0059597 |                37 |   1951x | 265.059 us | 3.47% | 256.315 us | 0.62% | 39.014G | 312.115 GB/s | 16.13% |   2031x | 251.270 us |
|     I32 |      U64 |          INSERT |          L2 |      REGS | 100000000 | 300000000 |         2 |       0.24194 |                37 |    217x |   2.316 ms | 0.39% |   2.307 ms | 0.04% | 43.341G | 346.728 GB/s | 17.92% |    227x |   2.302 ms |
|     I32 |      U64 |        CONTAINS |        GMEM |      GMEM |  10000000 | 300000000 |         2 |     0.0059597 |                37 |   1793x | 287.897 us | 3.22% | 278.999 us | 0.46% | 35.842G | 286.740 GB/s | 14.82% |   1906x | 262.407 us |
|     I32 |      U64 |        CONTAINS |        GMEM |      GMEM | 100000000 | 300000000 |         2 |       0.24194 |                37 |    192x |   2.621 ms | 0.64% |   2.612 ms | 0.54% | 38.282G | 306.254 GB/s | 15.82% |    199x |   2.617 ms |
|     I32 |      U64 |        CONTAINS |        GMEM |      REGS |  10000000 | 300000000 |         2 |     0.0059597 |                37 |   1831x | 282.127 us | 3.39% | 273.182 us | 0.85% | 36.606G | 292.845 GB/s | 15.13% |   1946x | 258.885 us |
|     I32 |      U64 |        CONTAINS |        GMEM |      REGS | 100000000 | 300000000 |         2 |       0.24194 |                37 |    197x |   2.552 ms | 0.38% |   2.543 ms | 0.13% | 39.323G | 314.587 GB/s | 16.25% |    207x |   2.528 ms |
|     I32 |      U64 |        CONTAINS |          L2 |      GMEM |  10000000 | 300000000 |         2 |     0.0059597 |                37 |   1873x | 276.017 us | 3.42% | 266.967 us | 0.43% | 37.458G | 299.662 GB/s | 15.48% |   1961x | 260.320 us |
|     I32 |      U64 |        CONTAINS |          L2 |      GMEM | 100000000 | 300000000 |         2 |       0.24194 |                37 |    194x |   2.593 ms | 0.72% |   2.584 ms | 0.63% | 38.706G | 309.651 GB/s | 16.00% |    196x |   2.599 ms |
|     I32 |      U64 |        CONTAINS |          L2 |      REGS |  10000000 | 300000000 |         2 |     0.0059597 |                37 |   1891x | 273.571 us | 3.51% | 264.540 us | 0.81% | 37.802G | 302.412 GB/s | 15.63% |   1939x | 258.965 us |
|     I32 |      U64 |        CONTAINS |          L2 |      REGS | 100000000 | 300000000 |         2 |       0.24194 |                37 |    198x |   2.543 ms | 0.36% |   2.534 ms | 0.06% | 39.458G | 315.667 GB/s | 16.31% |    204x |   2.529 ms |

File: include/cuco/detail/cache_residency_control.cuh [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+ */
+void unregister_l2_persistence(cudaStream_t& stream)
+{
+  cudaStreamAttrValue stream_attribute;
+  // Setting the window size to 0 to disable it
+  stream_attribute.accessPolicyWindow.num_bytes = 0;
+  // Overwrite the access policy attribute of CUDA Stream
+  cudaStreamSetAttribute(stream, cudaStreamAttributeAccessPolicyWindow, &stream_attribute);
+  // Remove any persistent lines in L2$
+  cudaCtxResetPersistingL2Cache();
```
Comment (sleeepyjack) [inline include/cuco/detail/cache_residency_control.cuh:87]: This resets the cache globally. Maybe undesirable.

File: include/cuco/detail/cache_residency_control.cuh [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  // Must be less than cudaDeviceProp::accessPolicyMaxWindowSize
+  auto const num_bytes = std::min(std::distance(begin, end) * sizeof(value_type),
+                                  std::size_t(prop.accessPolicyMaxWindowSize));
+
+  cudaDeviceSetLimit(cudaLimitPersistingL2CacheSize, carve_out * prop.persistingL2CacheMaxSize);
+
+  // Stream level attributes data structure
+  cudaStreamAttrValue stream_attribute;
+  // Global Memory data pointer
+  stream_attribute.accessPolicyWindow.base_ptr = reinterpret_cast<void*>(&begin[0]);
```
Comment (sleeepyjack) [inline include/cuco/detail/cache_residency_control.cuh:59]: Yes, this is indeed ugly

**Comment by kkraus14** [comment]
@sleeepyjack we would love to see this work pushed forward so we can utilize this. Is there anything that we can do to help here?

**Comment by sleeepyjack** [comment]
@kkraus14 I can move this up on my task list and hammer out a new draft PR tomorrow so we can get started on discussing the last few design questions. I'll keep you posted.

**Comment by sleeepyjack** [comment]
Superseeded by #573

**Pull request closed without merging**